lib/trento/activity_logging/parser/metadata_enricher.ex

```diff
@@ -0,0 +1,125 @@
+defmodule Trento.ActivityLog.Logger.Parser.MetadataEnricher do
+  @moduledoc """
+  Metadata enricher enriches metadata extracted by activity parser.
+  """
+
+  alias Trento.ActivityLog.ActivityCatalog
+
+  alias Trento.Clusters.Projections.ClusterReadModel
+  alias Trento.Databases.Projections.DatabaseReadModel
+  alias Trento.Hosts.Projections.HostReadModel
+  alias Trento.SapSystems.Projections.SapSystemReadModel
+
+  alias Trento.{Clusters, Databases, Hosts, SapSystems}
+
+  @spec enrich(activity :: ActivityCatalog.activity_type(), metadata :: map()) ::
+          {:ok, maybe_enriched_metadata :: map()}
+  def enrich(activity, metadata) do
+    case ActivityCatalog.detect_activity_category(activity) do
+      supported_activities
+      when supported_activities in [:connection_activity, :domain_event_activity] ->
+        {:ok, enrich_metadata(activity, metadata)}
+
+      _ ->
+        {:ok, metadata}
+    end
+  end
+
+  defp enrich_metadata(activity, metadata),
+    do:
+      {activity, metadata}
+      |> maybe_enrich_with_hostname()
+      |> maybe_enrich_with_cluster_name()
+      |> maybe_enrich_with_database_sid()
+      |> maybe_enrich_with_sap_system_sid()
+      |> elem(1)
+
+  defp maybe_enrich_with_hostname({activity, metadata}) do
+    enriched_metadata =
+      with {:ok, host_id} <- detect_enrichment(:host, {activity, metadata}),
+           {:ok, %HostReadModel{hostname: hostname}} <- Hosts.by_id(host_id) do
+        Map.put(metadata, :hostname, hostname)
+      else
+        _ -> metadata
+      end
+
+    {activity, enriched_metadata}
+  end
+
+  defp maybe_enrich_with_cluster_name({activity, metadata}) do
+    enriched_metadata =
+      with {:ok, cluster_id} <- detect_enrichment(:cluster, {activity, metadata}),
+           {:ok, %ClusterReadModel{name: cluster_name}} <- Clusters.by_id(cluster_id) do
+        Map.put(metadata, :name, cluster_name)
+      else
+        _ -> metadata
+      end
+
+    {activity, enriched_metadata}
+  end
+
+  defp maybe_enrich_with_database_sid({activity, metadata}) do
+    enriched_metadata =
+      with {:ok, database_id} <- detect_enrichment(:database, {activity, metadata}),
+           {:ok, %DatabaseReadModel{sid: sid}} <- Databases.by_id(database_id) do
+        Map.put(metadata, :sid, sid)
+      else
+        _ -> metadata
+      end
+
+    {activity, enriched_metadata}
+  end
+
+  defp maybe_enrich_with_sap_system_sid({activity, metadata}) do
+    enriched_metadata =
+      with {:ok, sap_system_id} <- detect_enrichment(:sap_system, {activity, metadata}),
+           {:ok, %SapSystemReadModel{sid: sid}} <- SapSystems.by_id(sap_system_id) do
+        Map.put(metadata, :sid, sid)
+      else
+        _ -> metadata
+      end
+
+    {activity, enriched_metadata}
+  end
```

Comment:
These functions are identical except for a parameter (i.e. one of host, database, sap system etc.). What are your thoughts about re-parameterizing/genericizing this function into something like `maybe_enrich({activity, metadata}, CtxModule, field_to_extract, entity_atom)` where CtxModule is `SapSystems | Databases`, field to extract is `:name | :sid ...`, entity_atom is `:database |  :sap_system ...`.

Reply:
I thought about that, thanks for helping me fighting my laziness :smile:  :see_no_evil: 

Here's the change 03f6d8c9dfa3001fede760fc8601c96df0ee02d6: if we're gonna be needing to let's say enrich with multiple values, we'll revisit what needed :+1: